gradio/upload.py

```diff
@@ -0,0 +1,173 @@
+import os
+import random
+import huggingface_hub
+import json
+
+
+def get_existing_configuration(readme_file):
+    with open(readme_file, "r") as f:
+        readme = f.read()
+    if not readme.strip().startswith("---"):
+        return None
+    readme = readme.strip()
+    if "---" not in readme[3:]:
+        return None
+    readme = readme.split("---")[1].strip()
+    config = {}
+    for line in readme.split("\n"):
+        if ":" not in line:
+            continue
+        key, value = line.split(":", 1)
+        config[key.strip()] = value.strip()
+    return config
+
+
+def add_configuration_to_readme(repo_directory, readme_file):
+    dir_name = os.path.basename(repo_directory)
+    app_name = input(f"Enter Spaces app name [{dir_name}]: ") or dir_name
+
+    app_file = None
+    for file in os.listdir(repo_directory):
+        file_path = os.path.join(repo_directory, file)
+        if (not os.path.isfile(file_path)) or (
+            not file.endswith(".py") and not file.endswith(".ipynb")
+        ):
+            continue
+
+        with open(file_path, "r", encoding="utf-8", errors="ignore") as f:
+            content = f.read()
+            if "import gradio" in content:
+                app_file = file
+                break
+
+    app_file = (
+        input(f"Enter Gradio app file {f'[{app_file}]' if app_file else ''}: ")
+        or app_file
+    )
+    if not app_file or not os.path.exists(app_file):
+        print("Failed to find Gradio app file.")
+        return
+    notebook_file = None
+    if app_file.endswith(".ipynb"):
+        notebook_file = app_file
+        app_file = "_" + notebook_file.replace(".ipynb", ".py")
+
+    EMOJI_SET = "🤯🤖🧠🐶👑💥🐮🎁🐙🦋"
+    default_emoji = random.choice(EMOJI_SET)
+    emoji = input(f"Enter Spaces Card emoji [{default_emoji}]: ") or default_emoji
+
+    COLOR_SET = ["red", "yellow", "green", "blue", "indigo", "purple", "pink", "gray"]
+    default_color = random.choice(COLOR_SET)
+    color = input(f"Enter Spaces Card color [{default_color}]: ") or default_color
+
+    readme_header = f"""---
+title: {app_name}
+app_file: {app_file}
+emoji: {emoji}
+colorFrom: {color}
+colorTo: {color}
+sdk: gradio
+{f'notebook_file: {notebook_file}' if notebook_file else ''}
+---
+"""
+    with open(readme_file, "r") as f:
+        readme = f.read()
+    with open(readme_file, "w") as f:
+        f.write(readme_header + readme)
+
+
+def notebook_to_script(notebook_path, script_path):
+    print(f"Converting {notebook_path} to {script_path}.")
```

Comment:
You might want to look at [`nbconvert`](https://nbconvert.readthedocs.io/en/latest/usage.html#convert-script) instead of rolling your own.

Reply:
reason why I didn't:
- didn't want to add a dependency
- want to use custom logic to convert !shell commands to subprocess calls